src/app/TimedHandler.cpp

```diff
@@ -0,0 +1,138 @@
+    if (mState == State::kExpectingTimedAction)
+    {
+        // We were just created; our caller should have done this only if it's
+        // dealing with a Timed Request message.
+        VerifyOrDie(aPayloadHeader.HasMessageType(MsgType::TimedRequest));
+        mState         = State::kReceivedTimedAction;
+        CHIP_ERROR err = HandleTimedRequestAction(aExchangeContext, aPayloadHeader, std::move(aPayload));
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(DataManagement, "Failed to parse Timed Request action: handler %p exchange " ChipLogFormatExchange, this,
+                         ChipLogValueExchange(aExchangeContext));
+            StatusResponse::SendStatusResponse(Status::InvalidAction, aExchangeContext, /* aExpectResponse = */ false);
+        }
+        return err;
+    }
+
+    if (mState == State::kExpectingFollowingAction)
+    {
+        if (System::SystemClock().GetMonotonicTimestamp() > mTimeLimit)
+        {
+            // Time is up.  Spec says to send UNSUPPORTED_ACCESS.
+            ChipLogError(DataManagement, "Timeout expired: handler %p exchange " ChipLogFormatExchange, this,
+                         ChipLogValueExchange(aExchangeContext));
+            return StatusResponse::SendStatusResponse(Status::UnsupportedAccess, aExchangeContext, /* aExpectResponse = */ false);
+        }
+
+        if (aPayloadHeader.HasMessageType(MsgType::InvokeCommandRequest))
+        {
+            auto * imEngine = InteractionModelEngine::GetInstance();
+            aExchangeContext->SetDelegate(imEngine);
+            ChipLogDetail(DataManagement, "Handing timed invoke to IM engine: handler %p exchange " ChipLogFormatExchange, this,
+                          ChipLogValueExchange(aExchangeContext));
+            imEngine->OnTimedInvoke(this, aExchangeContext, aPayloadHeader, std::move(aPayload));
+            return CHIP_NO_ERROR;
+        }
+
+        // TODO: Add handling of MsgType::WriteRequest here.
+    }
+
+    // Not an expected message.  Send an error response.  The exchange will
+    // close when we return.
+    ChipLogError(DataManagement, "Unexpected unknown message in tiemd interaction: handler %p exchange " ChipLogFormatExchange,
+                 this, ChipLogValueExchange(aExchangeContext));
+
+    return StatusResponse::SendStatusResponse(Status::InvalidAction, aExchangeContext, /* aExpectResponse = */ false);
+}
+
+void TimedHandler::OnExchangeClosing(Messaging::ExchangeContext *)
+{
+    InteractionModelEngine::GetInstance()->OnTimedInteractionFailed(this);
+}
+
+CHIP_ERROR TimedHandler::HandleTimedRequestAction(Messaging::ExchangeContext * aExchangeContext,
+                                                  const PayloadHeader & aPayloadHeader, System::PacketBufferHandle && aPayload)
+{
+    using namespace Protocols::InteractionModel;
+
+    constexpr TLV::Tag timeoutTag = TLV::ContextTag(0);
+
+    // Parse the action, store things, etc.
+    System::PacketBufferTLVReader reader;
```

Comment:
Also #if CHIP_CONFIG_IM_ENABLE_SCHEMA_CHECK
   ReturnErrorOnFailure(timedRequestParser.CheckSchemaValidity());
#endif

so that we can have pretty print in linux and darwin platform?

Reply:
Alright, I'll use TimedRequestMessage bits here.